Fix adding event to queue cache

igchor opened the pull request:
in single-device scenario.

When event is an internal event (when using queue with discard events property) the UrQueue is set to nullptr. This means that inside addEventToContextCache the event will be added to a multi-device cache which is wrong. This causes unbounded growth of the multi-device event cache if there are no multi-device events requested from the cache.

igchor commented:
This is draft for now as I'd like to add some tests for this.

pbalcer commented:
This is a regression after #1324. We should target this for 0.9.

Given the severity of the impact, it makes me wonder whether the discard events are used by anyone and worth keeping around in their current form.  They are adding a lot of complexity to the codebase.

igchor commented:
I've added one test for the event caches but it's a bit hacky (I'm using ZeCallCount map from L0 adapter) - if you have a better idea how to test number of zeEventCreateCalls then I'm open for suggestions.  The testcase that verifies the issue fixed by this PR is `eventsReuseNoVisibleEvent` with `UR_QUEUE_FLAG_DISCARD_EVENTS`.

Also, there are some problems I found:
- the test seems to fail with UR_RESULT_ERROR_FEATURE_UNSUPPORTED (this is returned from either urQueueFinish or urQeueuRelease) for some machines on our CI (with older driver?). For now, I've added a GTEST_SKIP() if this error is returned but it requires further investigation
- test does not work with `UR_QUEUE_FLAG_SUBMISSION_BATCHED` - I'm not sure why
- in eventsReuseNoVisibleEvent I would expect the number of `zeEventCreate` to never be as high as `numIters * numEnqueues`  but that number is equal or higher unless `UR_QUEUE_FLAG_DISCARD_EVENTS` is set and `UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE` is NOT set

pbalcer commented on test/adapters/level_zero/event_cache_tests.cpp:
```diff
@@ -0,0 +1,170 @@
+using FlagsTupleType = std::tuple<ur_queue_flags_t, ur_queue_flags_t,
+                                  ur_queue_flags_t, ur_queue_flags_t>;
+
+#define ASSERT_SUCCESS_OR_UNSUPPORTED(ret)                                     \
```
can this be moved to some common location?

pbalcer commented on test/adapters/level_zero/event_cache_tests.cpp:
```diff
@@ -0,0 +1,170 @@
+        for (int j = 0; j < numEnqueues; j++) {
+            enqueueWork(nullptr, i * numEnqueues + j);
+        }
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urQueueFinish(queue));
```
it's odd that the UNSUPPORTED is returned by urQueueFinish. Did you investigate what specifically is returning that?

pbalcer commented on source/adapters/level_zero/common.cpp:
```diff
@@ -88,7 +88,9 @@ bool setEnvVar(const char *name, const char *value) {
 
 // This will count the calls to Level-Zero
+#ifndef UR_L0_CALL_COUNT_IN_TESTS
 std::map<std::string, int> *ZeCallCount = nullptr;
```
I really don't like this. #1454 is meant to give programmatic access to internals like this in tests. So please add a TODO.

pbalcer commented on test/adapters/level_zero/event_cache_tests.cpp:
```diff
@@ -0,0 +1,170 @@
+    ::testing::Combine(
+        testing::Values(0, UR_QUEUE_FLAG_DISCARD_EVENTS),
+        testing::Values(0, UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE),
+        // TODO: why the test fails with UR_QUEUE_FLAG_SUBMISSION_BATCHED?
```
with what combination of the flags? The discard events path doesn't seem that well tested, so it's not inconceivable that's its broken when a few different flags are set.

igchor replied:
I rerun this on the latest driver and everything works fine there so I guess we can just uncomment this once we switch our CI to the newer driver.
